### PR TITLE
ENHANCEMENT Hide Image_Backend construction behind image manipulations to improve performance

### PR DESCRIPTION
**Pull request by tractorcow**
Fixes https://github.com/silverstripe/silverstripe-asset-admin/issues/536

Should be tested with https://github.com/silverstripe/silverstripe-assets/pull/100 for best effect

### CODE REVIEW

**Comment by tractorcow** [comment]
Yay fixed it. Just needed another warmCache() and it's better. :)

This whole PR makes the asset admin a lot faster!

File: src/ImageManipulation.php [hunk 1]
```diff
@@ -230,14 +231,16 @@ public function existingOnly()
...
-            return $this;
+        if (!$width || !$height || !is_numeric($width) || !is_numeric($height)) {
+            throw new InvalidArgumentException("Dimensions must be non empty integers");
         }
-
         $variant = $this->variantName(__FUNCTION__, $width, $height, $backgroundColor, $transparencyPercent);
         return $this->manipulateImage(
             $variant,
             function (Image_Backend $backend) use ($width, $height, $backgroundColor, $transparencyPercent) {
+                if ($backend->getWidth() == $width && $backend->getHeight() == $height) {
```
[the remark this answers is not in the document]
Comment (tractorcow) [reply src/ImageManipulation.php]: Values could be strings coming from templates, db, so I've not required it to be strictly equivalent.

**Comment by flamerohr** [comment]
nice one, I think `validateDimension()` is more appropriately named `castDimension()` or `convertDimension()`
I would have expected `validateSomething()` to return `true` or `false`

**Comment by tractorcow** [comment]
Renamed to `checkDimension` @flamerohr

**Comment by dhensby** [comment]
Hey team, this shouldn't have gone into a patch release given it's adding new APIs and appears to be more substantial than just a bug fix.

It's also introducing exceptions to code that previously worked (see #103) that are not going to be handled by existing user code. Not only this but these method calls with invalid parameters are coming from core code, so the core code should be handling those exceptions rather than passing responsibility on to the end dev.

I think either `castDimension` needs to be a lot more forgiving with the values it gets given or we need to cast our variables to `int` before passing them through to it (which kind of defeats the purpose).

**Comment by tractorcow** [comment]
Well, the idea is that a silent-ish error became a less silent error. The fact that floats incorrectly cause exceptions was an oversight rather than an intentional breakage. We'll fix by allowing numeric values other than ints.

**Comment by tractorcow** [comment]
> Also, the method feels badly named as it doesn't appear to cast the variable to an int, it throws exceptions if it isn't a suitable int-like var.

We had a huge discussion about the name that no one could really agree on... this was the best compromise we could come up with.

**Comment by dhensby** [comment]
> Well, the idea is that a silent-ish error became a less silent error. The fact that floats incorrectly cause exceptions was an oversight rather than an intentional breakage. We'll fix by allowing numeric values other than ints.

my point is that adding new exceptions in patch releases is unnecessarily risky and shouldn't be considered a "bug fix".

**Comment by tractorcow** [comment]
Oh well, we've fixed it now. :)

**Pull request merged**